### DISCUSSION TITLE
0.18.0: Functional Differential Geometry complete!

### DISCUSSION THREAD

**Discussion by sritchie** [sicmutils/sicmutils#345] [General]
This release focused on porting over all of the material required to run every piece of code from Sussman and Wisdom's ["Functional Differential Geometry"](http://xahlee.info/math/i/functional_geometry_2013_sussman_14322.pdf). The namespaces are lightly documented; the situation is better than the original library, but will only get better as I work through the material and add commentary.

There is a huge amount of functionality and material here! We can run many examples from general and special relativity, and the tests are full of exercises from the classic ["Gravitation" book by Misner, Thorne and Wheeler (MTW)](https://www.amazon.com/Gravitation-Charles-W-Misner/dp/0691177791).

Notable changes from the rest of the library:

- `Operator` instances are slightly more efficient with their addition and multiplication, handling `zero?` and `one?` cases appropriately

- `Structure`s can now hold metadata

- We've extended the SICMUtils generics to Clojure's Map and Set data structures. These can now combine with `+`. Maps are treated as sparse infinite-dimensional vector spaces, and can multiply with symbolic or numeric scalars.

- `ModInt` instances are now correctly equal to numbers (when those numbers mod down to the `ModInt` instance's residue).

### What's next?

The next major change will be an overhaul of the simplifier to make it work fast enough to solve Einstein's field equations in a reasonable amount of time, maybe even in the browser. Polynomial GCD is slow, but [#341](https://github.com/sicmutils/sicmutils/pull/341) will make it fast.

On to the detailed notes!

### Functional Differential Geometry

- From #339:

  - The new `sicmutils.calculus.covariant/Lie-D` can compute the Lie derivative for coordinates.

  - `sicmutils.calculus.frame` lets us create relativistic reference frames for investigating special relativity problems. This namespace aliases the following functions into `sicmutils.env`: 'frame?', `make-event`, `event?`, `claim`, `coords->event`, `event->coords`, `ancestor-frame`, `frame-name`, `frame-owner` and `frame-maker`.

  - `sicmutils.calculus.hodge-star` implements the Hodge star operator from chapter 10 of Functional Differential Geometry, plus Gram Schmidt orthonormalization. This namespace aliases the following functions into `sicmutils.env`: `Gram-Schmidt`, `orthonormalize` and `Hodge-star`.

  - `sicmutils.calculus.indexed` ports over the scmutils work on indexed objects and typed functions. This namespace aliases the following functions into `sicmutils.env`: `argument-types`, `with-argument-types`, `index-types`, `with-index-types`, `typed->indexed`, `indexed->typed`, `typed->structure`, `structure->typed`, `i:outer-product` and `i:contract`.

  - `sicmutils.calculus.manifold` gains `coordinate-system?`, which (predictably) returns true if its argument is a coordinate system, false otherwise. `chart` and `point` also take relativistic reference frames in addition to coordinate systems; the returned function converts to and from coordinates and events, rather than coordinates and manifold points.

  - `Div`, `Grad`, `Curl` and `Lap` move from `sicmutils.calculus.derivative` to `sicmutils.calculus.vector-calculus`. This namespace also contains versions of these operators from Functional Differential Geometry. This namespace aliases the following functions into `sicmutils.env`: `divergence`, `curl`, `gradient` and `Laplacian` (along with the others mentioned).

  - lots of new namespaces available in `sicmutils.env.sci`, soon to be deployed to Nextjournal: `sicmutils.calculus.{hodge-star, indexed, vector-calculus}`, and `sicmutils.sr.{boost,frames}`.

  - `sicmutils.sr.boost` describes boosts from special relativity, covered in chapter 11 of Functional Differential Geometry. This namespace aliases the following functions into `sicmutils.env`: `make-four-tuple`, `four-tuple->ct`, `four-tuple->space`, `proper-time-interval`, `proper-space-interval`, `general-boost`, `general-boost2` and `extended-rotation`.

  - `sicmutils.sr.frames` implements relativistic reference frames from special relativity, covered in chapter 11 of Functional Differential Geometry. This namespace aliases the following functions into `sicmutils.env`: `make-SR-coordinates`, `SR-coordinates?`, `SR-name`, `make-SR-frame`, `the-ether`, `boost-direction`, `v:c`, `coordinate-origin`, `add-v:cs` and `add-velocities`.

- From #338:

  - `sicmutils.fdg.bianchi-test` verifies the Bianchi identities; this was a challenge posed by GJS, and getting it working exposed a few bugs and triggered the rest of the work in this PR. Thank you, GJS!

  - `covariant-derivative` now properly handles the case of functions with argument types attached.

  - added `covariant-differential` to `sicmutils.calculus.covariant`.

  - aliased all functions from various namespaces in `sicmutils.calculus` into `sicmutils.env`.

  - adds `sicmutils.calculus.metric`, with the following functions exposed in `sicmutils.env`:

      - `coordinate-system->metric-components`, `coordinate-system->metric`, `coordinate-system->inverse-metric`, `literal-metric`, `components->metric`, `metric->components`, `metric->inverse-components`, `metric-over-map`, `lower`, `vector-field->oneform-field`, `drop1`, `raise`, `oneform-field->vector-field`, `raise1`, `drop2`, `raise2`, `trace2down`, `trace2up`, `sharpen`, `S2-metric`

      - `sicmutils.calculus.metric/invert` is exposed as `metric:invert` to match the scmutils naming scheme.

  - adds `sicmutils.calculus.connection`, with the following functions exposed in `sicmutils.env`:

    - `make-Christoffel-1`, `metric->Christoffel-1`, `metric->Christoffel-2`, `literal-Christoffel-1`, `literal-Christoffel-2`, `metric->connection-1`, `metric->connection-2`, `literal-Cartan`, `structure-constant`

- #337:

  - adds `sicmutils.calculus.curvature`, with these new functions and many tests from the classic "Gravitation" book: `Riemann-curvature`, `Riemann`, `Ricci`, `torsion-vector`, `torsion` and `curvature-components`

  - form fields now have NO identity operator, since they multiply by wedge, not
    composition.

- #328 adds many utilities for "Functional Differential Geometry".

  - vector fields, in `sicmutils.calculus.vector-field`:

    - new functions: `basis-components->vector-field`,
      `vector-field->basis-components`

    - vector fields now implement `v/zero?` and `v/zero-like` by returning
      proper vector fields.

  - form fields, in `sicmutils.calculus.vector-field`:

    - new functions: `nform-field?`, `basis-components->oneform-field`, `oneform-field->basis-components` and `function->oneform-field` (aliased as `differential-of-function`)

    - `Alt`, `alt-wedge` provide alternate wedge product definitions

    - form fields now implement `v/zero?` and `v/zero-like` by returning proper form fields that retain their rank.

    - form fields now correctly multiply via `*` by using `sicmutils.calculus.form-field/wedge`, instead of composition.

  - maps between manifolds, in `sicmutils.calculus.map`:

    - new function: `pushforward-function`

    - `differential` becomes `differential-of-map`, aliased back as `differential`

  - `sicmutils.calculus.covariant` gains new functions: `Cartan?`, `Christoffel?`, `Cartan->Christoffel`, `symmetrize-Christoffel`, `symmetrize-Cartan`, `Cartan->Cartan-over-map`, `geodesic-equation`, `parallel-transport-equation`.

  - `sicmutils.calculus.covariant/vector-field-Lie-derivative` can now handle structural inputs.

### New Functions, Functionality

- From #342:

  - Added `sicmutils.calculus.derivative/D-as-matrix` and `sicmutils.matrix/as-matrix`, ported from scmutils.

  - converted `sicmutils.modint.ModInt` to a `deftype`; this allows `ModInt` instances to be `=` to non-`ModInt` numbers on the right, if the right side is equal to the residue plus any integer multiple of the modulus. `v/=` gives us this behavior with numbers on the LEFT too, and `ModInt` on the right.

    - This change means that `:i` and `:m` won't return the residue and modulus anymore. `sicmutils.modint` gains new `residue` and `modulus` functions to access these attributes.

  - The JVM version of sicmutils gains more efficient `gcd` implementations for `Integer` and `Long` (in addition to the existing native `BigInteger` `gcd`), thanks to our existing Apache Commons-Math dependency.

  - `sicmutils.structure/dual-zero` aliases `compatible-zero` to match the scmutils interface. Both are now aliased into `sicmutils.env`.

  - `Structure` instances can now hold metadata (#339).

- From #339:

  - In `sicmutils.mechanics.rotation`:

    - gains aliases for `R{xyz}` in `rotate-x`, `rotate-y` and `rotate-z`.

    - `R{x,y,z}-matrix` now alias `rotate-{x,y,z}-matrix`.

    - Added new functions `angle-axis->rotation-matrix` and the mysterious,
      undocumented `wcross->w` from scmutils

    - `rotate-{x,y,z}-tuple` are now aliased into `sicmutils.env`.

  - `Operator` instances now ignore the right operator in operator-operator addition if the left operator passes a `v/zero?` test. Contexts are still appropriately merged.

  - in `sicmutils.simplify.rules`, the `sqrt-contract` ruleset now takes a simplifier argument and attempts to use it to simplify expressions internal to a square root. As an example, if two square roots in a product simplify to the same expression, we can drop the wrapping square root; otherwise multiplication is pushed under the root as before.

    - Added a missing rule in `simplify-square-roots` that handles roots of exponents with odd powers.

  - `sicmutils.matrix` changes:

    - `generate` has a new 2-arity version; if you supply a single dimension the returned matrix is square.

    - `diagonal?` returns true if its argument is a diagonal matrix, false
      otherwise.

  - A new namespace, `sicmutils.util.permute`:

    - `factorial` moved here from `sicmutils.generic`. It's still aliased into `sicmutils.env`.

    - new functions: `permutations`, `combinations`, `cartesian-product`, `list-interchanges`, `permutation-parity`, `permutation-interchanges`, `permute`, `sort-and-permute`, `subpermute`, `number-of-permutations`, `number-of-combinations`. See the tests for usage examples.

- From #338:

  - `(* <structure> <operator>)` multiplication pushes operator multiplication into the structure, rather than converting a structure into an operator.

- #337:

  - If you combine `Operator` instances with non-equal `:subtype` fields, the returned operator now keeps the parent subtype (or throws if one is not a subtype of the other).

  - `Operator` instances now ignore any `identity?`-passing operator on the left or right side of operator-operator multiplication. Contexts are still appropriately merged.

  - Similarly, `Operator` addition ignores `zero?` operators on the left or right side, and subtraction ignores `zero?` operators on the right right.

- #328:

  - Closes #249; operators now verify compatible contexts on multiplication.

  - `Operator` instances can now provides custom `zero?`, `one?`, `identity?`, `zero-like`, `one-like` and `identity-like` implementations by setting a function of a single (operator-typed) argument to a keyword like `:zero?` in their context. the identity operator returns `true` for `identity?`, and `false` for `one?` so that it isn't stripped by the `g/*` function.

  - structures implement the 0-arity case of IFn now.

- #335 implements `g/make-rectangular`, `g/make-polar` `g/real-part` and `g/imag-part` for clojure's Map data structure. Maps are treated as sparse vectors, any missing key on either side of `make-rectangular` or `make-polar`is treated as a 0 (rather than an error because the keys don't match, as in vectors).

- #334 adds implementations of `g/add` and the `sicmutils.value.Value` protocol for clojure's Set data structure. Addition is defined as set union, and `(zero-like <set>)` returns the empty set.

- #334 implements `g/add`, `g/negate` and `g/sub` for Clojure's Map data structure. Map addition is defined as a merge using `g/add` on clashing values; `g/sub` is the same, but any values on the right side not on the left side are negated.

  Maps can also be multiplied with scalars (commutatively) or divided (scalar on the right side only) by scalars. This, plus the commutative group property declared above, mean that Clojure's maps are sparse vector spaces over anything that responds true to `sicmutils.value/scalar?`... currently anything in the numeric tower up to complex, along with symbolic expressions and `Differential` instances.

<hr /><em>This discussion was created from the release <a href='https://github.com/sicmutils/sicmutils/releases/tag/v0.18.0'>0.18.0: Functional Different Geometry complete!</a>.</em>

**Comment by phasetr** [comment]
I have a question on functions for Functional Differential Geometry in sicmutils? In P.8, there exists a function `d/dt`, and the footnote 8 says that `(define-coordinates t R1-rect)` defines `d/dt`. But I do not find the function `define-coordinates`. Does this function have a different name in sicmutils? And what symbol does a function `d/dt` in sicmutils? (In Clojure, a slash `/` has a syntactic meaning.)

Furthrmore, when I have some questions about functions for Functional Differential Geometry, may I write question on this thread? Should I build another discussion thread?

**Reply by sritchie** [reply]
@phasetr hey, so glad you're writing! another thread is probably best, so we can start to build up a list of differences from the book implementation as you find them.

We don't currently have a `define-coordinates`, but I agree that this is a good thing to add for REPL and notebook usage, and book compatibility. What we DO have right now is a slightly tidier `let-coordinates`: https://github.com/sicmutils/sicmutils/blob/b953c904d22797f22363abac345b74dc4edeef55/src/sicmutils/calculus/coordinate.cljc#L80-L109

The docstring on that page has a good example of usage, and what gets bound inside of that form. There are LOTS of uses of `let-coordinates` in the tests too: https://github.com/sicmutils/sicmutils/search?q=let-coordinates

The answer for `d/dt` specifically is that, you're right, we can't use `/` since it has a specific syntactic meaning for vars. All (I think) slashes like this get bound to `:`, so inside `let-coordinates` we see bindings like `d:dt` or `d:dtheta`:

```clojure
(testing "From McQuistan: Scalar and Vector Fields, pp. 103-106"
  (let-coordinates [[r theta zeta] R3-cyl
                    [x y z] R3-rect]
    (let [A (+ (* 'A_r d:dr)
               (* 'A_theta d:dtheta)
               (* 'A_z d:dzeta))
          p ((m/point R3-rect) (up 'x 'y 'z))]
      (is (= '(up (/ (+ (* -1N A_theta y (sqrt (+ (expt x 2) (expt y 2))))
                        (* A_r x))
                     (sqrt (+ (expt x 2) (expt y 2))))
                  (/ (+ (* A_theta x (sqrt (+ (expt x 2) (expt y 2))))
                        (* A_r y))
                     (sqrt (+ (expt x 2) (expt y 2))))
                  A_z)
             (simplify
              ((vf/vector-field->components A R3-rect) (up 'x 'y 'z))))))))
```

Let me know if that helps, and please feel free to make a new thread for FDG stuff!

**Reply by phasetr** [reply]
@sritchie Thank you for your reply! I understand. I will list up conflicts between the book and sicmutils in a new thread.
